### DISCUSSION TITLE
Subcommand delimiter

### DISCUSSION THREAD

**Discussion by fdncred** [nushell/nushell#6460] [Polls]
We've been contemplating changing the subcommand delimiter, which is currently a space.

Here are some options we're considering.
Option 1:
```
# current (space)
str replace
```
Option 2:
```
# single colon
str:replace
```
Option 3:
```
# double colon
str::replace
```
Option 4:
```
# single period
str.replace
```

**Comment by Zoybean** [comment]
My primary preference is "not space", because I think there's value in clearly distinguishing parameters from subcommands, especially if nu keeps not needing $ for variable access.
As for what non-space delimiter, I'm not sure if `.` could lead to confusion with table access, so I'm leaning toward `:` and `::`

**Comment by generic-github-user** [comment]
If the semantics of subcommands and fields (?) in records line up closely enough I think a period makes sense and would be the most elegant option. It is a little less ergonomic, but I definitely agree with @Zoybean regarding the separation between parameters and subcommands - though requiring a colon for interactive use is probably further than I would be willing to bend to maintain that consistency.

**Reply by Zoybean** [reply]
What do you mean by "requiring a colon for interactive use"? I feel like I've missed something

**Reply by generic-github-user** [reply]
Sorry, I should've clarified; I mean that using `:` or `::` as a command delimiter makes sense for scripting but feels somewhat too unergonomic for a command shell. I think leaving it as an option would be reasonable as long as a more efficient way is still available.

**Comment by cschierig** [comment]
I personally am against using anything other than spaces as subcommand delimiters because of backwards-compatibility with "normal" programs. If I only had to use nu plugins in my shell, it wouldn't be a problem because I would learn to write `into.bool`, but there are many programs which don't have and some which will never have a nu plugin alternative. when scripting or typing commands, I then would have to remember if the command I'm using is internal or external. 
Especially when using command chains, it would be very annoying to type or read:
`echo "5day" | into.duration | foo subfoo | bar.x --full`
`subfoo` is a subcommand of the external command `foo`, but to me it now looks more like an argument passed to it because the internal commands `into` and `bar` use `.` as a delimiter.

**Comment by pepdiz** [comment]
AFAIK a subcommand is coneptually a parameter to the command, so space is the right separator.

If a more explicit marking of subcommads is desired I think the period is a good option assuming the paralelism between subcommands and fields. If explicit marking is required maybe would be better using :  to be explicit saying it's a subcommand and not a field.

The pros of keep using space are:

- backwards compatibility
- clarity  and readability ( it's clearer to write `into int` than `into.int`)
- introspection, inside a definition you can access the subcommand as a parameter rather to turst in specifix accesor methods

The (maybe) cons:

- maybe confusion in definitions with long parameters   (i.e.  mycmd mysubcmd -flag param x y z)  
- no explicit mark for subcommands and thus no explicit difference between subcommands and parameters

**Comment by theAkito** [comment]
I think, it should remain space.

The reason for that is, that it feels more natural & I inituitively immediately got the point, when I saw it the first time.

The worst alternative of all would be `str::replace`. That's the worst. Please, don't do that.
The second worst is `str:replace`.
While `str.replace` would barely be acceptable, it still doesn't feel right.
`str replace` is just right. It's Nu. It's *Nu*tural or perhaps *Nu*taral. 😄

It also has already very good core support, like e.g. `def "str myStr" [] {}`, etc. So, no need to change it.

Finally, note that space is the only thing that everyone gets immediately.
The more symbols you introduce, the more difficult it becomes to grasp the language.
Just look at Ruby or something, where you learn symbols instead of the language...
While colons aren't that bad, they are the first step in the wrong direction.

P.S.: Using space is the quickest to type, of all proposed methods. When programming, you don't have to be the fastest, but when using a Shell you want to be able to type quickly, especially in a shell like Nu, where you have to type a lot to process data, etc.
